candle-core/src/metal_backend.rs

```diff
@@ -70,16 +112,107 @@ impl MetalDevice {
         &self.device
     }
 
-    pub fn new_buffer(&self, element_count: usize, dtype: DType) -> Buffer {
+    pub fn new_buffer(&self, element_count: usize, dtype: DType) -> Arc<Buffer> {
         let size = (element_count * dtype.size_in_bytes()) as NSUInteger;
-        self.device
-            .new_buffer(size, MTLResourceOptions::StorageModeManaged)
+        self._new_buffer(size, MTLResourceOptions::StorageModePrivate)
+    }
+
+    fn _new_buffer(&self, size: NSUInteger, option: MTLResourceOptions) -> Arc<Buffer> {
+        let mut buffers = self.buffers.try_write().unwrap();
```

Comment:
When would this `unwrap` fails? Feels a bit that this function should return a `Result` and propagate the error properly just in case

Reply:
It never should, at least in single threaded, otherwise it's a pretty bad bug in the scheduling/allocator.

In multi threaded, we need to decide, if we want to wait on the lock (with a timeout probably since deadlocks could really happen).
I think I'm ok delaying this decision when the current implem for single threaded as proven it's worth (I don't think there's a lot of benefit of doing multithreaded encoding of the commands at the moment).

Moved to Result though.